Blush integration

twagner9 opened the pull request:
# Description

Blush regularization, a method that involved the creation of a convolutional neural network (CNN), was created for use within RELION-5 around a year ago. The idea is to iteratively pass blocks of density maps to the CNN to de-noise the volume to improve 3D refinement of structures.

The full paper can be found here: 
[Kimanius, D., Jamali, K., Wilkinson, M.E. et al. Data-driven regularization lowers the size barrier of cryo-EM structure determination. Nat Methods 21, 1216–1221 (2024). https://doi.org/10.1038/s41592-024-02304-8](https://www.nature.com/articles/s41592-024-02304-8)

The original code was written in Python, using the PyTorch library for creating and executing the CNN, and executing in cisTEM required setting up a Python environment, installing dependencies, and calling to a bash script to execute the code and pass the density map to Python.

This solution re-creates the model using the LibTorch C++ API, and proceeds to load weights from the Python model into the C++ model, conducts all necessary pre- and post-processing steps on the model before and after passage to the model, including proper blocking behavior, within a standalone CLI program (see blush_refinement.cpp) and merge3d.cpp. It is completed in a way that allows the user using the cisTEM GUI to simply indicate whether they wish Blush to be applied or not. Thus, everything is run natively in C++.

However, the current solution uses **dynamic** linking, which requires that users have the CPU-only LibTorch version 2.5.0 installed and properly linked for their system. It is a bit challenging to use this on older operating systems, as the differences in kernel often mean that it is required that the user compile the LibTorch C++ API from source so that their system is capable of running it.

I have attempted to generate a statically built version to avoid this issue, but ran into many challenges as the PyTorch/LibTorch devs do not encourage static builds. I cannot quite recall at this time why that was the case.

For testing purposes, I was able to easily mount the necessary LibTorch version to the container.

Fixes # (issue)

# I have rebased my feature branch to be current with the master branch using to minimize conflicts and headaches

- [x] yes
- [ ] no

# Which compilers were tested

- [ ] g++
- [x] icpc
- [ ] clang
- [ ] other (please specify)

# These changes are isolated to the

- [x] gui
- [ ] core library
- [ ] gpu core library
- [x] program it modifies

# How has the functionality been tested?

Please describe the tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [x] Tested manually from GUI
- [x] Tested manually from CLI
- [ ] Passed console tests
- [ ] Passed samples functional testing
- [ ] other (please specify)

# Checklist:

- [ ] I have not changed anything that did not need to be changed
- [x] I have performed a self-review of my own code
- [x] I have commented my code, (***w.r.t. why***), particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the [documentation](https://github.com/bHimes/cisTEM_docs) {Ok to pass for now}
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

twagner9 commented:
This may fail the automated tests due to the LibTorch library not being present in the test environments. Also, I am definitely interested to know of any improvements that could be in the handling of the configure.ac and Makefile as it pertains to the new blush checks. What I have works, but it seems as though there might be better than options than what I currently do, like the copying and removal of the weights file.

bHimes commented:
maybe you thought of yours, but did you try building with static linking except for libtorch? Their licensing is bsd-3 (permissive) so we could bundle the dynamic libs with the cistem binaries and use rpath on build to bias the runtime linker to our included .so

i've had to play games like this, similarly setting LDRUNPATH to spoof the intel compiler into linking a function that overrides AMD checks that hurt performance

bHimes commented on configure.ac:
```diff
@@ -383,6 +383,56 @@ AC_ARG_ENABLE(openmp, AS_HELP_STRING([--enable-openmp],[Compile with OpenMP thre
   CXXFLAGS="$CXXFLAGS $OPENMP_CXXFLAGS"
   fi
   ])
+
```
Let's move this to its own m4 module. Configure AC is already a bit bloated. See the fastFFT submodule for a general idea.

bHimes commented on src/Makefile.am:
```diff
@@ -850,8 +871,8 @@ project3d_LIBTOOLFLAGS = $(LIBTOOL_FLAGS)
 refine3d_CXXFLAGS = $(WX_CPPFLAGS_BASE)
-refine3d_CPPFLAGS = $(WX_CPPFLAGS_BASE)
-refine3d_LDADD    = libcore.a $(WX_LIBS_BASE) $(MKL_LIBS)
+refine3d_CPPFLAGS = $(WX_CPPFLAGS_BASE) 
```
Any idea why there are new whitespaces here?

bHimes commented on src/Makefile.am:
```diff
@@ -1108,6 +1129,14 @@ azimuthal_average_CPPFLAGS = $(WX_CPPFLAGS_BASE)
+if ENABLE_BLUSH_REFINEMENT_AM
+blush_refinement_SOURCES  = programs/blush_refinement/blush_refinement.cpp programs/blush_refinement/blush_model.cpp
+blush_refinement_CXXFLAGS = $(WX_CPPFLAGS_BASE)
+blush_refinement_CPPFLAGS = $(WX_CPPFLAGS_BASE) $(BLUSH_CPPFLAGS) -diag-disable 1011 -diag-disable 2196 -diag-disable 1628 -diag-disable 186
```
These flags are breaking the non-icc builds in CI. They should be added in config.ac (well, in the new submodule that is referenced from config.ac) and only added when the compiler is icc. Also, a comment on why they are disabled would be good.

twagner9 replied:
Ah yes, I was aware of this when I commented them out, but had never gotten around to making them compiler specific.

Will add a comment; many warnings that are only occurring because of LibTorch flood the compiler output (return statement in void function, missing return statement, etc.) so I will note this.

bHimes reviewed (changes requested):
Made a few comments about things to fix while we think about how to build/distribute. Let's iterate on these changes and then revisit that bigger question.

twagner9 committed:
Squash commits for integrated blush into cisTEM

twagner9 committed:
Squash commits for integrated blush into cisTEM

twagner9 commented on src/programs/merge3d/merge3d.cpp:
```diff
@@ -1,5 +1,32 @@
+// within defines.h; they must be undefined to be able
+// to compile successfully with LibTorch.
+#ifdef BLUSH
+#ifdef NONE
```
I have rebased my branch and was utilizing the libtorch helper, which works nicely. I do want to get your thoughts on something though.

Some files may need multiple headers from the LibTorch library, and these (from my experimentation with it) also needed the LibTorch specific macros for the types. So, to enable the ability to still use this simpler, less error-prone format of having a header, I broke apart the push/undef/pop logic in `cistem_torch_helper.h`, placing the the push/undef in `include/libtorch/libtorch_push_macros.h` and the pop for restoring the cisTEM macros in `include/libtorch/libtorch_pop_macros.h`.

This enables the ability to enclose multiple headers that are specific to a needed LibTorch operation between the `libtorch_push_macros.h` and `libtorch_pop_macros.h`.

I'm therefore wondering if this seems reasonable, or if there's an alternative preferred solution. I did play around with making a macro and using _Pragma but it seems that icpc's pre-processor is quite old and didn't support what I was trying.

I will be pushing this code relatively soon so you can take a look.

bHimes replied:
So that makes enough sense for now. Really though, we should be defining these as class scoped enums in constants.h 

Absolutely okay with keeping this as a TODO issue, but should probably fix sooner than later. It will touch quite a bit of code, so I'm happy to do it post Oct review since these are a bunch of pretty big PRs

twagner9 commented on src/programs/blush_refinement/blush_weights.dat:
Will remove.

I wasn't sure how to approach this in a portable way, which was why I ended up leaving the binary .dat (also mostly for functional/testing purposes -- I knew there would be some things to fix when I PR'd this branch). I tried serializing it at one point but was left with the same issue of size, and past that I was more focused on ensuring everything worked.

twagner9 committed:
rebase onto environment, incorporating some recs

-Rebased onto #560 to take advantage of some of the dev environment changes.
-Added submodule_Blush.m4 for conditional compilation of LibTorch/Blush routines.
-configure.ac: moved libtorch/blush conditions to submodule_Blush. Added submodule_Blush to configure.ac.
-Split cistem_torch_helper.h logic into two headers to allow for multiple includes from torch headers.
-refine3d.cpp: WIP; put blush in the refinement rather than the merge logic -- may give better results, and spares users from having to run generate 3D on a volume to get the non-blushed form.
-merge3d.cpp: remove blush logic
-GUI Panels: shifting blush around to the refine3d logic rather than merge3d; includes arguments changes in the IPC via the JobPackage class.
-Files in src/programs/blush_refinement: added the new format for splitting the logic for avoiding macro conflicts.
-console_test.cpp: added new format for splitting the logic for avoiding macro conflicts.
-Deleted blush_weights.dat

twagner9 committed:
rebase onto environment, incorporating some recs

-Rebased onto #560 to take advantage of some of the dev environment changes.
-Added submodule_Blush.m4 for conditional compilation of LibTorch/Blush routines.
-configure.ac: moved libtorch/blush conditions to submodule_Blush. Added submodule_Blush to configure.ac.
-Split cistem_torch_helper.h logic into two headers to allow for multiple includes from torch headers.
-refine3d.cpp: WIP; put blush in the refinement rather than the merge logic -- may give better results, and spares users from having to run generate 3D on a volume to get the non-blushed form.
-merge3d.cpp: remove blush logic
-GUI Panels: shifting blush around to the refine3d logic rather than merge3d; includes arguments changes in the IPC via the JobPackage class.
-Files in src/programs/blush_refinement: added the new format for splitting the logic for avoiding macro conflicts.
-console_test.cpp: added new format for splitting the logic for avoiding macro conflicts.
-Deleted blush_weights.dat

twagner9 commented:
@bHimes 

Pushed my latest changes.

I still haven't incorporated quite everything yet. After discussion with @timothygrant80 it was decided to move the blush logic from merge3d.cpp to refine3d.cpp. This might give better results, and should ensure the user will not have to execute Generate 3D to be able to see the volume without blush actively applied in the final stage. These changes are why it took me so long to start incorporating suggestions, so apologies for that!

This is still a WIP, as I still have to determine the best way to handle masking (blush relies on a radial mask to determine which blocks to pass to the model, but we also apply a mask/automask before starting refinement), but this is less relevant than the codebase issues.

bHimes commented:
> @bHimes
> 
> Pushed my latest changes.
> 
> I still haven't incorporated quite everything yet. After discussion with @timothygrant80 it was decided to move the blush logic from merge3d.cpp to refine3d.cpp. This might give better results, and should ensure the user will not have to execute Generate 3D to be able to see the volume without blush actively applied in the final stage. These changes are why it took me so long to start incorporating suggestions, so apologies for that!
> 
> This is still a WIP, as I still have to determine the best way to handle masking (blush relies on a radial mask to determine which blocks to pass to the model, but we also apply a mask/automask before starting refinement), but this is less relevant than the codebase issues.

I can see where that makes sense. The regularization is mostly for refinement, not for the final product, right?

twagner9 commented:
@bHimes 

On the topic of the blush_weights.dat binary, I was trying to think of some alternative ways to handle the weights since they shouldn't be in the main repo.

### 1. Automated during compile

I was thinking that the weights could be hosted within a tar archive somewhere (on GitHub, perhaps, if we make a releases page, or cistem.org since that is where releases currently exist) and then conditionally downloaded via `curl` and extracted automatically during the compile process based on whether blush is enabled by writing a make rule associated with the program that uses it.

### 2. Download during config

Basically the same as compile, but instead of making a rule in the makefile, the rule would be in configure.ac -- or more likely the blush module.

### 3. Create a script that must be run manually.

Unlike the above 2 options, this is somewhat less automated as it will require the user/dev to know the script exists to be able to download the weights. The fact that the script exists could be conveyed at the end configuration in a print stipulating that `download_blush_weights.sh` must be executed in order for blush to be applied, but this would require some other downstream warnings/checks to be issued to prevent runtime failure if the user misses it which is more time invested in making sure the user is downloading the needed weights.

I have thought a bit about alternative options, like converting to an array that can be stored in a C array that can then be directly embedded in the binary, but the size required starts to outweigh the advantage, which seems to be true of any options that involve embedding these weights.

The common theme for the download -> extract options are that the weights need to be hosted somewhere, and for the first 2, just devising a way to conditionally check if the weights are already present to avoid re-downloading during each configuration or compile process, but that should be relatively straightforward.

Let me know what you think of these, or if you have additional ideas!

>I can see where that makes sense. The regularization is mostly for refinement, not for the final product, right?

Yes, definitely better in refine3d for that reason.

bHimes commented:
In the current setup for distribution of compiled code we already are including the binary library .so in the build dir right? Why not just treat the weights like the library?

twagner9 committed:
Added masking service

Added MaskingService to unify the logic associated with masking in any of the 3D panels. Gives centralized location to make updates to the masking logic and avoid "oops, forgot to change 1 of 3 panels' masking logic".

Fixed some bugs with the RefineCTF panel and the call to masking thread.

Created BlushHelpers.cpp that has an ApplyBlush function in its namespace that will always  be what is called when blush inference is being performed. Avoids the issue of placing blush inference in one of the 3d refinement progs, which wastes computation time for each copy that is doing a redundant inference.

Added option for user to specify the number of threads to use for blush inference to all 3D refinement panels, capped at 128.

Left some commented out logic in the event that it needs to be returned to. These will be removed in the final commit.

twagner9 committed:
Batched blush

-Added batching to reduce the number of forward passes to the blush model, drastically improving inference runtime for large volumes.

-Added GNU CPU only config and build options to tasks.json.

-Fixed masking bugs -- chiefly the capping of resolution during autorefinement based on class_high_res_limits, and updating the refinement manager's current reference filenames.

-Added check to ensure that volume size can fit within the blush model block size of 64 pixels and its stride size of 20 pixels.

-Changed messaging with cancellation of blush to be done directly by panel rather than worker thread.

-Removed blush flags from binaries that don't need them in the Makefile.

twagner9 committed:
Remove debugging holdover used for simple, manual profiling.

twagner9 committed:
Fix macro conditionals

Swithced ENABLE_BLUSH macros in code to cisTEM_USING_BLUSH and removed unnecessary #undefs that are now isolated, once more, to the push/pop macro headers.

twagner9 committed:
Cleanup

Removing unnecessary comments, debugging comments/code, adding comment context for configuration and LibTorch related code.

Fixed masking bug wherein current reference filenames were not updated after masking/blush for the 3D refinement panels.

Corrected unnecessary code changes that would affect other logic.

Removed several additions to Makefile.am that are now optional programs.

twagner9 committed:
try adding enable libtorch to pass CI
